File: vokimi_frontend/src/ts/page_classes/manage_test_page/test_statistics/templates_shared/TestStatisticsTestTakenRecordsCount.ts
export class TestStatisticsTestTakenRecordsCount {
    readonly all: number;
    readonly loggedUp: number;
    readonly byFollowers: number;
    readonly byFriends: number;
    readonly lastHour: number;
    readonly lastDay: number;
    readonly lastMonth: number;
    readonly lastYear: number;
    //dynamic
    constructor(
        all: number,
        loggedUp: number,
        byFollowers: number,
        byFriends: number,
        lastHour: number,
        lastDay: number,
        lastMonth: number,
        lastYear: number
    ) {
        this.all = all;
        this.loggedUp = loggedUp;
        this.byFollowers = byFollowers;
        this.byFriends = byFriends;
        this.lastHour = lastHour;
        this.lastDay = lastDay;
        this.lastMonth = lastMonth;
        this.lastYear = lastYear;
    }
}

File: vokimi_api/Src/dtos/responses/manage_test_page/statistics/templates_shared/TestStatisticsTestTakenRecordsCount.cs
using vokimi_api.Src.db_related.db_entities;
using vokimi_api.Src.db_related.db_entities.users;
using vokimi_api.Src.db_related.db_entities_ids;

namespace vokimi_api.Src.dtos.responses.manage_test_page.statistics.templates_shared
{
    public record class TestStatisticsTestTakenRecordsCount(
        int All,
        int ByFollowers,
        int ByFriends,
        int LastHour,
        int LastDay,
        int LastMonth,
        int LastYear
    )
    {
        public static TestStatisticsTestTakenRecordsCount Create(
            ICollection<BaseTestTakenRecord> testTakings,
            HashSet<AppUserId> creatorsFollowers,
            HashSet<AppUserId> creatorFriends
        ) => new(
            testTakings.Count(),
            testTakings.Where(x => x.UserId is not null && creatorsFollowers.Contains(x.UserId.Value)).Count(),
            testTakings.Where(x => x.UserId is not null && creatorFriends.Contains(x.UserId.Value)).Count(),
            testTakings.Count(x => x.Date > DateTime.Now - TimeSpan.FromHours(1)),
            testTakings.Count(x => x.Date > DateTime.Now - TimeSpan.FromDays(1)),
            testTakings.Count(x => x.Date > DateTime.Now - TimeSpan.FromDays(30)),
            testTakings.Count(x => x.Date > DateTime.Now - TimeSpan.FromDays(365))
        );
    }
}
